File: src/jobs.mjs
import { fileURLToPath } from "node:url";
import path from "node:path";
import { spawnDetached, withLock, commitAll, sync } from "./git.mjs";
import { writeIndex } from "./index-file.mjs";
const BIN = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "bin", "memory.mjs");
// The write-completion job (index regeneration, commit, push), run either in
// this process (tests, and the detached child's own entry point) or spawned
// as a separate detached process. The one place both paths funnel through,
// so a caller of afterWrite never needs its own MEMORY_SYNC_INLINE branch.
export function runJobInline(bundle, dirRel, message) {
  withLock(bundle.root, () => {
    const dir = dirRel ? bundle.dir(dirRel.replace(/^projects\//, "")) : bundle.dir(null);
    writeIndex(bundle, dir);
    commitAll(bundle.root, message);
    sync(bundle.root, { pull: false, push: true });
  });
}
export function afterWrite(bundle, { dirRel, message }) {
  if (process.env.MEMORY_SYNC_INLINE === "1") { runJobInline(bundle, dirRel, message); return; } // tests run the job inline instead
  spawnDetached([BIN, "_job", "--dir", bundle.root, "--index", dirRel, "--commit", message]);
}

File: memory-go/jobs.go
package memory

import "strings"

// Job is the write-completion work Node runs in a detached process after a
// concept write: regenerate the directory's index, commit everything and
// push. A zero Job runs nothing. The caller decides whether to run it in a
// goroutine; the SDK never starts one.
type Job struct {
	DirRel  string
	Message string
}

// Run executes the job under the git lock. ran is false when the lock was
// held (the next write's job commits everything pending) or the job is zero.
func (j Job) Run(b *Bundle) (bool, error) {
	if j.Message == "" {
		return false, nil
	}
	return WithLock(b.Root, "git", lockStale, func() error {
		dir, err := b.Dir(strings.TrimPrefix(j.DirRel, "projects/"))
		if err != nil {
			return err
		}
		if err := b.WriteIndex(dir); err != nil {
			return err
		}
		if _, err := CommitAll(b.Root, j.Message); err != nil {
			return err
		}
		_, err = SyncGit(b.Root, false, true)
		return err
	})
}
